Allow node label to be a flexible expression

cpoole opened the pull request:
This PR addresses a new feature:

I do not see a way to provide a list of labels that can be OR'd together to select a node. Right now if you provide multiple labels it must match ALL of them. 

This PR makes node labels into a user specifiable expression:

`(node_labels['region'] == 'us-west-2' && node_labels['app'] == 'nginx') || (node_labels['region'] == 'us-west-2' && node_labels['foo'] == 'bar') || (node_labels['type'] == 'toolbox')`

This is achieved through the use of https://github.com/antonmedv/expr

~This PR adds the ability to flip draino into OR mode.~

~While doing this I realized when multiple labels are supplied that have the same key but different values only the last provided is used in the app: eg. `--node-label=foo=bar --node-label=foo=elephant` results in the `*nodeLabels` map only including `{foo: elephant}`~

~In order to fix this I parse them as an array of independent `map[string]string` and then let the Filter functions determine how to handle the deduping (required in the AND case to maintain backwards compatible behavior)~

dbenque commented:
I have the same problem but your solution would block me in fact because I need a mix of AND and OR.
I was about to start something where the flag can be repeated (OR) and in each repetition we can have a set of KEY=PAIR comma separated (AND).
This is possible because label value can't have a comma. The validation regexp for value is `(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?`

So for example:
`--node-label=foo=bar,app=nginx  --node-label=foo=bar,app=mysql    --node-label=type=toolbox`
would mean:
`( foo=bar AND app=nginx) OR (foo=bar,app=mysql) OR (type=toolbox)`

cpoole commented:
hah, I literally started with that! 

But then i realized it would be a breaking change. People that are currently specifying `node-label` more than once are expecting the AND behavior. 

I'm fine to tweak my PR to work the way you're proposing but maybe we should do a new flag? `--node-label-v2`

jacobstr commented:
The `--node-labels` map roughly to the `matchLabels` behavior. Started going down the path of thinking about e.g. `--node-match-expressions`:

https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements

```
--node-match-expressions="{key: node_type, operator: In, values: [ingress,control,worker]}"
```

Thats a rough angle I thought to offer up. I think, realistically, that the yaml encoded string encompassed in the command-line arg is kind of ugly.

I think in general, if folks are starting to specify more complex selectors a configuration file may start to be appropriate. The structure for these selectors would then mirror the kubernetes resources: leveraging data structures users are already familiar with as well as library functions in client-go.

Thus far, the answer has been to just use multiple draino instances for things like this, but that's not very user-centric admittedly. 

On the other hand, my personal feeling is that the command line arguments are best suited for some happy path scenario.

---

I appreciate the effort for backwards compatibility here as well.  I think I like the comma separated syntax more, but don't like the break in compatibility.

Why not keep the AND'ing behavior between `--node-labels` and use to indicate the `OR'd` alternatives?

jacobstr commented:
Sounds like I'm repeating myself with the configFile suggestion: https://github.com/planetlabs/draino/issues/62#issuecomment-564173998

dbenque commented:
Having the config file is clearly the way to more complex scenarii. I will probably work on it at some point.

Waiting for that a simple alternative can be to have 2 flags that are exclusive:
--node-label     <--  current flag, with current behavior
--node-labels   <--  new flag with comma separated option

repeating --node-label would be an AND
repeating --node-labels would be an OR between list of labels
any tentative to use both syntax at a time would be an error.

WDYT?

jacobstr commented:
Wanted to draw attention to this. Why not just allow:

```
--node-label worker_type=ingress,worker --node-label region=us-west1
```

Which would evaluate as:

```
The node must be (worker_type=ingress OR worker_type=worker) AND (region=us-west1)
```

The comma separation in the `--node-label` is in a sense † new and therefor non breaking.

@dbenque your suggestion was to use commas for `AND`, I'm suggest using `commas` to indicate `OR`. I don't think one is more intuitive than the other. As long as the behavior is:

- Clearly documented.
- Backward compatible.

We should be g2g.

† Note: Possibly someone in the wild is using a comma literally, but I think that's highly unlikely and might be disallowed anyway by the valid chars in labels values.
- [ ] Check that indeed one cannot created labels with commas in the value.

cpoole commented:
@jacobstr I agree a configfile is a better end state here. But I dont have the time for a refactor of that magnitude. 

What you're proposing would be fine with me if the syntax was `--node-label key=value,key=value`. I think many people will need the OR condition to work for different label keys. 

Who here can actually approve a PR/design? I'm not trying to put more work into this if a design cannot be solidified. I've currently forked and am running this branch in production. 

@negz you have the collaborator flair, but also @jacobstr you seem like you have merge access?

jacobstr commented:
@cpoole I work at Planet and hold the keys these days.  I think we're iterating on a good design here through the discussion so this debate isn't without merit.

Agreed re: `keyFoo=value,keyBar=value` for broadening the utility.

cpoole commented:
@jacobstr ok sounds good to me

So I'll go ahead and implement it using the following:

- each independent `--node-labels` collection functions as an AND
- within a `--node-labels` arg the multiple labels are OR'd together.

logic example:
```--node-labels foo=bar,animal=dog --node-labels region=us-west-2```

the following nodes would _match_:
Node1: 
foo=bar
region=us-west-2

Node2:
animal=dog
region=us-west-2

the following nodes would _not match_:
Node1:
foo=no
region=us-west-2

Node2:
foo=bar
region=us-west-1

dbenque commented:
@cpoole @jacobstr , maybe a stupid question but with that semantic how would I express:
```
(region=us-west-1 AND app=nginx) OR (region=us-west-2 AND foo=bar) OR (type=toolbox)
```

jacobstr commented:
I'm rusty on my boolean algebra buuuut... I the gist of what you'd be trying to do is converting something that it's in [disjunctive normal form](https://en.wikipedia.org/wiki/Disjunctive_normal_form) (or of ands) to [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) (and of ors). 

I believe this is generally possible, though a bit messy and perhaps re-expressing things this way will obfuscate the original meaning of the expression.

Some comments here: https://math.stackexchange.com/questions/1710210/converting-formula-from-cnf-to-dnf

dbenque commented:
Such transformation (dnf/cnf) would make it completely obscure while with a simple OR there is a natural way of expressing group of labels that define the group of nodes.
I am afraid the syntax proposed here https://github.com/planetlabs/draino/pull/75#issuecomment-616872086 would make it impossible to use for me and my users.

cpoole commented:
ok @dbenque and @jacobstr 

I'm finally circling back to this. I thought about what what @dbenque was saying and I agree... the original proposal in this PR was too restrictive. I've redone the PR using https://github.com/antonmedv/expr as an expression compiler. 

This allows for incredibly flexible use cases and I think will satisfy 99% of users. 

However, This is now a fully breaking change... What would y'all like to do?

jacobstr commented:
I hadn't looked at the push script before and was hoping that we didn't publish the latest tag - might have made me bolder here. I suppose the land shifting and breaking beneath your feet is to be expected for folks relying on latest. Unfortunately, I don't have a mailing list of draino users to even attempt to provide the courtesy of a heads up here.

I'll find some time to review things. I suspect we might have some way to put this new syntax behind a new command line flag and have the old-style flag "converted" to your expression compiler to reduce duplication. That'd let us avoid any migration mess.

jacobstr commented on internal/kubernetes/nodefilters.go:
```diff
@@ -17,28 +17,48 @@ and limitations under the License.
+
+		parameters := make(map[string]interface{}, 8)
+
+		parameters["node_labels"] = nodeLabels
```
I could see someone coming along with a use-case where anything in `metadata` is fair game. Heck, why not any part of the spec? :trollface:  So I don't think that needs to be implemented up front, but doing it in a forwards compatible manner would it be possible to have:

```
parameters["metadata"]["labels"] = nodeLabels
```

I threw this up in a go playground: https://play.golang.org/p/jrrQfyOH5P7

The goal would be that as use-cases come up for okaylisting parts of the node spec, there'd be an intuitive mapping between the shape of `kubectl get node <node> -o yaml` and our syntax.

jacobstr commented on cmd/draino/draino.go:
```diff
@@ -172,8 +173,18 @@ func main() {
+	var nodeLabelFilter cache.ResourceEventHandler
+
+	log.Sugar().Infof("label expression: %v", nodeLabelsExpr)
+	nodeLabelFilterFunc, err := kubernetes.NewNodeLabelFilter(nodeLabelsExpr)
```
It feels like we could retain nodeLabels by e.g. using the expression builder to convert the old-style syntax to the new. I hacked together a little demo: https://play.golang.org/p/rgS2rPm1SEP

jacobstr commented:
@cpoole I'm going to check in around 9:30-3:00 PST tomorrow to drive this to a conclusion. Happy to get this merged vs continuing to linger in the bike shed.

cpoole commented:
Sounds good @jacobstr I'll address your comments tonight

cpoole commented on internal/kubernetes/nodefilters_test.go:
```diff
@@ -21,15 +21,17 @@ import (
 	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 func TestNodeLabelFilter(t *testing.T) {
```
added the old ones back as a separate test

cpoole commented on internal/kubernetes/nodefilters.go:
```diff
@@ -17,28 +17,50 @@ and limitations under the License.
+
+		expression, err := expr.Compile(*expressionStr, expr.Env(parameters))
+		if err != nil {
+			log.Fatalf("Could not compile the node label expression: %v", err)
```
went ahead and compiled outside of the body... seems like I was planning to do that anyway since I changed the signature to return an err, but i must have gotten distracted :laughing:

cpoole commented:
@jacobstr addressed your comments. The build is failing because linter is running out of memory. I dont see a way to retry a travis build. Any suggestions?

jacobstr commented:
I'll merge [this](https://github.com/planetlabs/draino/pull/88) in a moment assuming tests pass and a subsequent rebase on your branch should hopefully resolve it.

cpoole commented:
@jacobstr squashed and rebased. Lets put this saga to an end

cpoole commented:
@jacobstr bump
